core/src/test/java/org/apache/iceberg/TestSequenceNumberForV2Table.java

```diff
@@ -0,0 +1,328 @@
+  }
+
+  @Test
+  public void testSingleTransaction() {
+    Transaction txn = table.newTransaction();
+    txn.newAppend().appendFile(FILE_A).commit();
+    txn.commitTransaction();
+    Assert.assertEquals("Snapshot sequence number should match expected",
+        1, table.currentSnapshot().sequenceNumber());
+  }
+
+  @Test
+  public void testConcurrentTransaction() {
+    Transaction txn1 = table.newTransaction();
+    Transaction txn2 = table.newTransaction();
+    Transaction txn3 = table.newTransaction();
+    Transaction txn4 = table.newTransaction();
+
+    txn1.newFastAppend().appendFile(FILE_A).commit();
+    txn3.newOverwrite().addFile(FILE_C).commit();
+    txn4.newDelete().deleteFile(FILE_A).commit();
+    txn2.newAppend().appendFile(FILE_B).commit();
+
+    txn1.commitTransaction();
+    txn2.commitTransaction();
+    txn3.commitTransaction();
+    txn4.commitTransaction();
+
+    Assert.assertEquals("Snapshot sequence number should match expected",
+        4, table.currentSnapshot().sequenceNumber());
+    validateDataFiles(files(FILE_A, FILE_B, FILE_C), seqs(4, 2, 3));
+  }
+
+  @Test
+  public void testMultipleOperationsTransaction() {
+    Transaction txn = table.newTransaction();
+    txn.newAppend().appendFile(FILE_A).commit();
+    Set<DataFile> toAddFiles = new HashSet<>();
+    Set<DataFile> toDeleteFiles = new HashSet<>();
+    toAddFiles.add(FILE_B);
+    toDeleteFiles.add(FILE_A);
+    txn.newRewrite().rewriteFiles(toDeleteFiles, toAddFiles).commit();
+    txn.commitTransaction();
+
+    Assert.assertEquals("Snapshot sequence number should match expected",
+        2, table.currentSnapshot().sequenceNumber());
+    validateDataFiles(files(FILE_A, FILE_B), seqs(2, 2));
+  }
+
+  @Test
+  public void testExpirationInTransaction() {
+    table.newFastAppend().appendFile(FILE_A).commit();
+    long snapshotId = table.currentSnapshot().snapshotId();
+    table.newAppend().appendFile(FILE_B).commit();
+
+    Transaction txn = table.newTransaction();
+    txn.expireSnapshots().expireSnapshotId(snapshotId).commit();
+    txn.commitTransaction();
+
+    Assert.assertEquals("Snapshot sequence number should match expected",
+        2, table.currentSnapshot().sequenceNumber());
+    validateDataFiles(files(FILE_B), seqs(2));
+  }
+
+  @Test
+  public void testTransactionFailure() {
+    table.newAppend()
+        .appendFile(FILE_A)
+        .appendFile(FILE_B)
+        .commit();
+
+    table.updateProperties()
+        .set(TableProperties.COMMIT_NUM_RETRIES, "0")
+        .commit();
+
+    table.ops().failCommits(1);
+
+    Transaction txn = table.newTransaction();
+    txn.newAppend().appendFile(FILE_C).commit();
+
+    AssertHelpers.assertThrows("Transaction commit should fail",
+        CommitFailedException.class, "Injected failure", txn::commitTransaction);
+
+    Assert.assertEquals("Snapshot sequence number should match expected",
+        1, table.currentSnapshot().sequenceNumber());
```

Comment:
This assertion validates that the table's current snapshot didn't change, not that the table did not assign as sequence number to the failed transaction. This should check `TableMetadata.lastSequenceNumber` instead.

Reply:
OK.